### ISSUE TITLE
Don't std::exit() on error

### ISSUE THREAD

**Issue by NovA80** [myint/optparse#2]
In the current implementation the `optparse` just stops the whole program on any parsing error (as well as after processing --help and --version options). This may be unacceptable, when some cleanup code should be executed before exiting, in parallel MPI environment, etc.

I propose to implement some more mild way to handle parser errors. I think about 3 possibilities:
- throw an exception instead of exiting
- set some internal flag e.g. `bool optparse::is_valid` that can be checked explicitly
- replace any calls of `std::exit()` by `optparse::exit()` method and make it virtual to allow overriding.

**Comment by myint** [comment]
I probably would have done it as an exception if I implemented this from scratch. I think the original author's (Johannes) intent was to match the Python `optparse` as closely as possible. But I think this is something worth adding. I suppose it would have to be an optional feature if we want to preserve backward compatibility.

**Comment by NovA80** [comment]
If we really want "to match the Python `optparse` as closely as possible" then according to https://docs.python.org/3/library/optparse.html#how-optparse-handles-errors we 
> need to subclass OptionParser and override its exit() and/or error() methods.

To support this, first of all the direct calls of `std::exit()` should be replaced by `optparse::exit()`. Such a minimal change will be fully backward compatible.

Then the exceptions may be introduced by providing an alternative class `optparse_e` inherited from `optparse` with the `exit()` overridden appropriately. Definitely this can be done by users themselves, but personally I would like to see such a class on stock.

**Comment by myint** [comment]
That seems like a good approach to me!

**Comment by NovA80** [comment]
I've created the pull request with the first minimal change.

As for the second change, which would introduce exceptions instead of `std::exit()`, there were doubts. On errors the call of `exit(2)` can be quite naturally replaced by `throw std::logic_error(msg)`. But I think it's logically wrong to throw any exception on *successful* `--help` & `--version`. I can't find simple universal method to resolve this...

Although, if we do not bother about "best practices" for C++ exceptions, we can just `throw 2;` or `throw 0;` as the upstream did in https://github.com/weisslj/cpp-optparse/pull/6 . As for me, such an ungraceful technique can be useful and I can commit this later in the form of subclass `OptionParserExcept`.

**Comment by NovA80** [comment]
Ok, the `OptionParserExcept` class has been merged. Thanks! It's really useful in my case. But there is one inconvenience with it. The shortended code
 ```
OptionParserExcept parser = OptionParserExcept()
          .version(version)
          .description("Description")
          .epilog("Epilog");
```
doesn't compile with the error
> error: conversion from ‘optparse::OptionParser’ to non-scalar type ‘optparse::OptionParserExcept’ requested.

The simple workaround is to declare empty object `OptionParserExcept parser;` and then apply chain of assigning methods `parser.version("vers").description("descr");`. This is ok, but it would be nice to have the same workflow as with the base class `OptionParser`. Unfortunately, I can't find a solution for now and I wonder if it possible. :)

**Comment by myint** [comment]
I see what you mean. To make it work it almost seems like you would have to override all those functions to make use of covariant return types. An alternative approach may be needed.
